ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -973,40 +1081,107 @@ private void handleMetadataBlob(BlobData blobData) throws IOException, MessageFo
+                          "Handling decrypt job call back for Metadata chunk {} to set decrypt callback results",
+                          blobId);
+                      decryptCallbackResultInfo.setResultAndException(result, exception);
+                      routerCallback.onPollReady();
+                    }));
+          } else {
+            initializeDataChunks();
           }
         }
       }
     }
 
+    /**
+     * Initialize data chunks and few other cast for metadata chunk
+     */
+    private void initializeDataChunks() {
+      chunkIdIterator = keys.listIterator();
+      numChunksTotal = keys.size();
+      dataChunks = new GetChunk[Math.min(keys.size(), NonBlockingRouter.MAX_IN_MEM_CHUNKS)];
+      for (int i = 0; i < dataChunks.length; i++) {
+        dataChunks[i] = new GetChunk(chunkIdIterator.nextIndex(), (BlobId) chunkIdIterator.next());
+      }
+    }
+
     /**
      * Process a simple blob and extract the requested data from the blob.
      * @param blobData the simple blob's data
+     * @param userMetadata userMetadata of the blob
+     * @param encryptionKey encryption key for the blob. Could be null for non encrypted blob.
      */
-    private void handleSimpleBlob(BlobData blobData) {
+    private void handleSimpleBlob(BlobData blobData, byte[] userMetadata, ByteBuffer encryptionKey) {
       totalSize = blobData.getSize();
       chunkSize = totalSize;
       boolean rangeResolutionFailure = false;
-      try {
+      if (encryptionKey == null) {
+        rangeResolutionFailure = resolveRange(totalSize);
+      } else {
+        // for encrypted blobs, Blob data will not have the right size. Will have to wait until decryption is complete
+      }
+      /*try {
         if (options != null && options.getBlobOptions.getRange() != null) {
-          resolvedByteRange = options.getBlobOptions.getRange().toResolvedByteRange(totalSize);
+          resolvedByteRange = options.getBlobOptions.getRange().toResolvedByteRange(blobProperties.getBlobSize());
         }
       } catch (IllegalArgumentException e) {
         onInvalidRange(e);
         rangeResolutionFailure = true;
-      }
+      }*/
       if (!rangeResolutionFailure) {
         chunkIdIterator = null;
         dataChunks = null;
         chunkIndex = 0;
         numChunksTotal = 1;
-        chunkIndexToBuffer.put(0, filterChunkToRange(blobData));
-        numChunksRetrieved = 1;
+        if (encryptionKey == null) {
+          chunkIndexToBuffer.put(0, filterChunkToRange(blobData.getStream().getByteBuffer()));
+          numChunksRetrieved = 1;
+        } else {
+          logger.trace("Submitting decrypt job for simple blob {}", blobId);
+          long startTimeMs = System.currentTimeMillis();
+          decryptCallbackResultInfo = new DecryptCallBackResultInfo();
+          progressTracker.initializeDecryptionTracker();
+          cryptoJobHandler.submitJob(new DecryptJob(blobId, encryptionKey, blobData.getStream().getByteBuffer(),
```

Comment:
just curious - why is `DecryptJob()` instantiated at three different places in code (I would have thought there will be at most two - one for data chunks and one for metadata chunk)

Reply:
data chunk, metadata chunk and simple chunk